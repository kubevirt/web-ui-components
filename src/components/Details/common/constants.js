export const DESCRIPTION_KEY = 'description';
export const FLAVOR_KEY = 'flavor';
