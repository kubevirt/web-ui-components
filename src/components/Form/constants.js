export const TEXT = 'text';
export const TEXT_AREA = 'textarea';
export const DROPDOWN = 'dropdown';
export const CHECKBOX = 'checkbox';
export const POSITIVE_NUMBER = 'positive-number';
export const LABEL = 'label';
export const CUSTOM = 'custom';
