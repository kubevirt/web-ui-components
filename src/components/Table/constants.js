export const ON_ACTIVATE = 'onActivate';
export const ON_CONFIRM = 'onConfirm';
export const ON_CANCEL = 'onCancel';
export const ON_CHANGE = 'onChange';
export const ON_MOVE = 'onMove';
export const ON_DELETE = 'onDelete';

export const ACTIONS_TYPE = 'actions';
export const DELETE_ACTION = 'delete';
