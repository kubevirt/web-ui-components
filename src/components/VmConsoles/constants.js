export const NIC_KEY = 'nic';
