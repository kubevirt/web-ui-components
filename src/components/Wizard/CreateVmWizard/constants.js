export const BASIC_SETTINGS_TAB_IDX = 0;
export const NETWORK_TAB_IDX = 1;
export const STORAGE_TAB_IDX = 2;
export const RESULTS_TAB_IDX = 3;

export const ALL_TABS = [BASIC_SETTINGS_TAB_IDX, NETWORK_TAB_IDX, STORAGE_TAB_IDX, RESULTS_TAB_IDX];

// BasicSettingsTab
export const NAME_KEY = 'name';
export const NAMESPACE_KEY = 'namespace';
export const DESCRIPTION_KEY = 'description';
export const IMAGE_SOURCE_TYPE_KEY = 'imageSourceType';
export const REGISTRY_IMAGE_KEY = 'registryImage';
export const IMAGE_URL_KEY = 'imageURL';
export const USER_TEMPLATE_KEY = 'userTemplate';
export const OPERATING_SYSTEM_KEY = 'operatingSystem';
export const FLAVOR_KEY = 'flavor';
export const MEMORY_KEY = 'memory';
export const CPU_KEY = 'cpu';
export const WORKLOAD_PROFILE_KEY = 'workloadProfile';
export const START_VM_KEY = 'startVM';
export const CREATE_TEMPLATE_KEY = 'createTemplate ';
export const CLOUD_INIT_KEY = 'cloudInit';
export const HOST_NAME_KEY = 'hostname';
export const AUTHKEYS_KEY = 'authKeys';
