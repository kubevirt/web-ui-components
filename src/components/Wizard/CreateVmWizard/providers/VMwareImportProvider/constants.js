export const PROVIDER_VMWARE = 'VMware';

export const PROVIDER_VMWARE_VCENTER_KEY = 'vCenterInstance';
export const PROVIDER_VMWARE_HOSTNAME_KEY = 'vmwareHostname';
export const PROVIDER_VMWARE_USER_NAME_KEY = 'vmwareUserName';
export const PROVIDER_VMWARE_USER_PASSWORD_KEY = 'vmwarePassword';
export const PROVIDER_VMWARE_REMEMBER_PASSWORD_KEY = 'rememberVMwareCredentials';

export const PROVIDER_VMWARE_CHECK_CONNECTION_KEY = 'checkConnectionButton';
export const PROVIDER_VMWARE_CHECK_CONNECTION_BTN_TEXT_KEY = 'checkConnectionButtonText';
export const PROVIDER_VMWARE_CHECK_CONNECTION_BTN_SAVE = 'Check and Save';
export const PROVIDER_VMWARE_CHECK_CONNECTION_BTN_DONT_SAVE = 'Check';
export const PROVIDER_VMWARE_STATUS_KEY = 'vmwareStatus';

export const PROVIDER_VMWARE_VM_KEY = 'vmwareVm';

export const PROVIDER_VMWARE_V2V_NAME_KEY = 'v2vVmwareName';
export const PROVIDER_VMWARE_NEW_VCENTER_NAME_KEY = 'newVCenterName';

export const PROVIDER_VMWARE_V2V_LAST_ERROR = 'PROVIDER_VMWARE_V2V_LAST_ERROR';
