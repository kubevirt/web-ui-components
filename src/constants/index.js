import { fedora28 } from '../k8s/mock_templates/fedora28.template';
import { rhel75 } from '../k8s/mock_templates/rhel75.template';
import { ubuntu1804 } from '../k8s/mock_templates/ubuntu1804.template';
import { rhelHighPerformance } from '../k8s/mock_templates/rhel-high-p.template';
import { windows } from '../k8s/mock_templates/windows.template';
import { linuxUserTemplate } from '../k8s/mock_user_templates/linux.template';
import { windowsUserTemplate } from '../k8s/mock_user_templates/windows.template';
import { network1 } from '../k8s/mock_network/network.template';
import { network2 } from '../k8s/mock_network/network2.template';

export const ANNOTATION_DEFAULT_DISK = 'defaults.template.cnv.io/disk';
export const ANNOTATION_DEFAULT_NETWORK = 'defaults.template.cnv.io/network';
export const ANNOTATION_FIRST_BOOT = 'cnv.ui.firstBoot';
export const ANNOTATION_PXE_INTERFACE = 'cnv.ui.pxeInterface';
export const ANNOTATION_USED_TEMPLATE = 'template.cnv.ui';

export const CLOUDINIT_DISK = 'cloudinitdisk';
export const CLOUDINIT_NOCLOUD = 'cloudInitNoCloud';
export const CLOUDINIT_VOLUME = 'cloudinitvolume';
export const CUSTOM_FLAVOR = 'Custom';
export const FLAVOR_LABEL = 'kubevirt.io/flavor';
export const KUBEVIRT_API_VERSION = 'kubevirt.io/v1alpha2';
export const OS_LABEL = 'kubevirt.io/os';
export const POD_NETWORK = 'Pod Networking';

export const PROVISION_SOURCE_PXE = 'PXE';
export const PROVISION_SOURCE_REGISTRY = 'Registry';
export const PROVISION_SOURCE_TEMPLATE = 'Template';
export const PROVISION_SOURCE_URL = 'URL';

export const PVC_ACCESSMODE_RWO = 'ReadWriteOnce';

export const REGISTRY_DISK = 'registrydisk';
export const REGISTRY_VOLUME = 'registryvolume';

export const TEMPLATE_API_VERSION = 'template.openshift.io/v1';
export const TEMPLATE_FLAVOR_LABEL = 'flavor.template.cnv.io';
export const TEMPLATE_OS_LABEL = 'os.template.cnv.io';
export const TEMPLATE_PARAM_VM_NAME = 'NAME';
export const TEMPLATE_TYPE_LABEL = 'template.cnv.io/type';
export const TEMPLATE_TYPE_VM = 'vm';
export const TEMPLATE_TYPE_BASE = 'base';
export const TEMPLATE_WORKLOAD_LABEL = 'workload.template.cnv.io';

export const VIRTIO_BUS = 'virtio';

export const baseTemplates = [fedora28, ubuntu1804, rhel75, rhelHighPerformance, windows];
export const networkConfigs = [network1, network2];
export const userTemplates = [linuxUserTemplate, windowsUserTemplate];
export const templates = [...baseTemplates, ...userTemplates];
