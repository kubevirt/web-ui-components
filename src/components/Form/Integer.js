import React from 'react';
import PropTypes from 'prop-types';
import { FormControl, noop } from 'patternfly-react';

import { eventValueHandler } from './utils';
import { getSequence, setNativeValue } from '../../utils/utils';
import { isMinus, KEY_CODES, INPUT_NAVIGATION_KEYS } from '../../constants/keys';

const NON_NEGATIVE_INTEGER_KEYS = [
  ...INPUT_NAVIGATION_KEYS,
  ...getSequence(KEY_CODES[0], KEY_CODES[9]),
  ...getSequence(KEY_CODES.NUMPAD[0], KEY_CODES.NUMPAD[9]),
];

const INTEGER_KEYS = [...NON_NEGATIVE_INTEGER_KEYS, KEY_CODES.MINUS, KEY_CODES.HYPHEN_MINUS, KEY_CODES.NUMPAD.SUBTRACT];

const POSITIVE_INT_REGEX = /^[1-9]\d*$/;
const NON_NEGATIVE_INT_REGEX = /^(0|[1-9]\d*)$/;
const INT_REGEX = /^(-?0|-?[1-9]\d*)$/;

const PRECEEDING_ZEROS_POSITIVE_INT_REGEX = /^0*([1-9]\d*)$/;
const PRECEEDING_ZEROS_INT_REGEX = /^(-?)0*([1-9]\d*)$/;

const fixPrecedingZerosPositiveInt = value => {
  const match = PRECEEDING_ZEROS_POSITIVE_INT_REGEX.exec(value);
  return match && match.length === 2 ? match[1] : '';
};

const fixPrecedingZerosNonNegativeInt = value => {
  const match = PRECEEDING_ZEROS_POSITIVE_INT_REGEX.exec(value);
  return match && match.length === 2 ? match[1] : '0';
};

const fixInt = (value, oldValue, keyCode) => {
  if (value.length === 0 && isMinus(keyCode)) {
    if (oldValue) {
      return oldValue < 0 || oldValue === '-0' ? oldValue : `-${oldValue}`;
    }
    return '-0';
  }

  const match = PRECEEDING_ZEROS_INT_REGEX.exec(value);
  if (match && match.length === 3) {
    if (match[1]) {
      return `${match[1]}${match[2]}`;
    }
    return match[2];
  }
  return '0';
};

const isInputValid = (allowedKeys, keyCode, targetValue, additionalValidation) => {
  if (allowedKeys.includes(keyCode)) {
    return additionalValidation ? additionalValidation(keyCode, targetValue) : true;
  }
  return false;
};

export const Integer = ({ id, value, disabled, defaultValue, onChange, onBlur, positive, nonNegative, className }) => {
  let allowedKeys;
  let validRegex;
  let fixAfterValue;
  let min;
  let additionalValidation;

  if (positive) {
    allowedKeys = NON_NEGATIVE_INTEGER_KEYS;
    validRegex = POSITIVE_INT_REGEX;
    fixAfterValue = fixPrecedingZerosPositiveInt;
    min = 1;
    additionalValidation = (keyCode, targetValue) =>
      !(targetValue === '' && (keyCode === KEY_CODES[0] || keyCode === KEY_CODES.NUMPAD[0]));
  } else if (nonNegative) {
    allowedKeys = NON_NEGATIVE_INTEGER_KEYS;
    validRegex = NON_NEGATIVE_INT_REGEX;
    fixAfterValue = fixPrecedingZerosNonNegativeInt;
    min = 0;
  } else {
    allowedKeys = INTEGER_KEYS;
    validRegex = INT_REGEX;
    fixAfterValue = fixInt;
  }

  const onKeydown = e => {
    const { keyCode, target } = e;

    if (!isInputValid(allowedKeys, keyCode, target.value, additionalValidation)) {
      e.preventDefault();
      return false;
    }

    const oldValue = target.value;

    window.setTimeout(() => {
      if (!validRegex.test(target.value)) {
        const v = fixAfterValue(target.value, oldValue, keyCode);
        setNativeValue(target, v);
        target.dispatchEvent(
          new Event('input', {
            bubbles: true,
            cancelable: true,
          })
        );
      }
    }, 0);
    return true;
  };

  return (
    <FormControl
      id={id}
      type="number"
      onKeyDown={onKeydown}
      min={min}
      value={value}
      defaultValue={defaultValue}
      onBlur={eventValueHandler(onBlur)}
      onChange={eventValueHandler(onChange)}
      className={className}
      disabled={disabled}
    />
  );
};

Integer.defaultProps = {
  value: undefined,
  defaultValue: undefined,
  onChange: noop,
  onBlur: noop,
  positive: false,
  nonNegative: false, // is ignored when positive == true
  className: undefined,
  disabled: false,
};

Integer.propTypes = {
  id: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  positive: PropTypes.bool,
  nonNegative: PropTypes.bool,
  className: PropTypes.string,
  disabled: PropTypes.bool,
};
