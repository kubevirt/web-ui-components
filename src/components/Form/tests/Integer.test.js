import React from 'react';
import { shallow } from 'enzyme';
import { FormControl, KEY_CODES as PF_KEY_CODES } from 'patternfly-react';
import { Integer } from '../Integer';
import { KEY_CODES } from '../../../constants/keys';

const testInteger = (value = -5, positive = false, nonNegative = false) => (
  <Integer id="1" value={value} onChange={jest.fn()} positive={positive} nonNegative={nonNegative} />
);

const getEvent = keyCode => ({
  keyCode,
  preventDefault: jest.fn(),
  target: {
    value: '',
  },
});

const testValidity = (component, keyCode, validity) => {
  const event = getEvent(keyCode);
  expect(
    component
      .find(FormControl)
      .props()
      .onKeyDown(event)
  ).toBe(validity);

  if (validity) {
    expect(event.preventDefault).not.toHaveBeenCalled();
  } else {
    expect(event.preventDefault).toHaveBeenCalled();
  }
};

const testValidKey = (component, keyCode) => testValidity(component, keyCode, true);

const testInValidKey = (component, keyCode) => testValidity(component, keyCode, false);

describe('<Integer />', () => {
  it('renders correctly', () => {
    const component = shallow(testInteger());
    expect(component).toMatchSnapshot();
  });

  it('renders positive correctly', () => {
    const component = shallow(testInteger(5, true));
    expect(component).toMatchSnapshot();
  });

  it('renders nonNegative correctly', () => {
    const component = shallow(testInteger(0, false, true));
    expect(component).toMatchSnapshot();
  });

  it('integer onKeyDown', () => {
    const component = shallow(testInteger());

    testValidKey(component, KEY_CODES['0']);
    testValidKey(component, KEY_CODES.NUMPAD['1']);
    testValidKey(component, KEY_CODES.NUMPAD.SUBTRACT);
    testValidKey(component, KEY_CODES.HYPHEN_MINUS);
    testValidKey(component, KEY_CODES.MINUS);
    testValidKey(component, KEY_CODES.LEFT_KEY);
    testValidKey(component, KEY_CODES.SHIFT);
    testValidKey(component, KEY_CODES.BACKSPACE);
    testInValidKey(component, PF_KEY_CODES.A);
  });

  it('nonNegative integer onKeyDown', () => {
    const component = shallow(testInteger(0, false, true));

    testValidKey(component, KEY_CODES['0']);
    testValidKey(component, KEY_CODES.NUMPAD['1']);
    testInValidKey(component, KEY_CODES.NUMPAD.SUBTRACT);
    testInValidKey(component, KEY_CODES.HYPHEN_MINUS);
    testInValidKey(component, KEY_CODES.MINUS);
    testValidKey(component, KEY_CODES.LEFT_KEY);
    testValidKey(component, KEY_CODES.SHIFT);
    testValidKey(component, KEY_CODES.BACKSPACE);
    testInValidKey(component, PF_KEY_CODES.A);
  });

  it('positive integer onKeyDown', () => {
    const component = shallow(testInteger(1, true));

    testValidKey(component, KEY_CODES['0']); // is set in timeout after that

    testValidKey(component, KEY_CODES.NUMPAD['1']);
    testInValidKey(component, KEY_CODES.NUMPAD.SUBTRACT);
    testInValidKey(component, KEY_CODES.HYPHEN_MINUS);
    testInValidKey(component, KEY_CODES.MINUS);
    testValidKey(component, KEY_CODES.LEFT_KEY);
    testValidKey(component, KEY_CODES.SHIFT);
    testValidKey(component, KEY_CODES.BACKSPACE);
    testInValidKey(component, PF_KEY_CODES.A);
  });
});
