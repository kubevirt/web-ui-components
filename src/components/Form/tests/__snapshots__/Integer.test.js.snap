// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Integer /> renders correctly 1`] = `
<FormControl
  bsClass="form-control"
  componentClass="input"
  disabled={false}
  id="1"
  onBlur={[Function]}
  onChange={[Function]}
  onKeyDown={[Function]}
  type="number"
  value={-5}
/>
`;

exports[`<Integer /> renders nonNegative correctly 1`] = `
<FormControl
  bsClass="form-control"
  componentClass="input"
  disabled={false}
  id="1"
  min={0}
  onBlur={[Function]}
  onChange={[Function]}
  onKeyDown={[Function]}
  type="number"
  value={0}
/>
`;

exports[`<Integer /> renders positive correctly 1`] = `
<FormControl
  bsClass="form-control"
  componentClass="input"
  disabled={false}
  id="1"
  min={1}
  onBlur={[Function]}
  onChange={[Function]}
  onKeyDown={[Function]}
  type="number"
  value={5}
/>
`;
